datafusion/physical-expr/src/expressions/cast.rs

```diff
@@ -132,6 +133,12 @@ impl PhysicalExpr for CastExpr {
             interval.cast_to(&cast_type, &self.cast_options)?,
         )])
     }
+
+    fn dyn_hash(&self, _state: &mut dyn Hasher) {
+        // `self.cast_options` doesn't support hashing
+        // Hence we cannot calculate `dyn_hash` for this type.
+        unimplemented!();
```

Comment:
I filed https://github.com/apache/arrow-rs/pull/4395 to add this upstream


Rather than `unimplmented` which will panic if this code is ever run (for eample if someone tries to put a CastExpr into a `HashSet`),  I think it is ok to hash the other fields while we wait for upstream.


The requirements on hash https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq are 

> In other words, if two keys are equal, their hashes must also be equal.

Which I believe this implementation would satisfy (though it will have some more hash collisions than when it would include `cast_options`):

```rust
impl Hash for CastExpr {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.expr.hash(state);
        self.cast_type.hash(state);
        // Add self.cast_options when hash is available
        // https://github.com/apache/arrow-rs/pull/4395
    }
}
```